### ISSUE TITLE
Stop using SHA-1 in eduPersonTargetedID generation

### ISSUE THREAD

**Issue by jaimeperez** [simplesamlphp/simplesamlphp#646]
Currently, `core:TargetedID` uses SHA1 to compute the `eduPersonTargetedID` attribute corresponding to a given user. Part of the input to the hash function is _user provided_ (the attribute configured as the user identifier), meaning the weaknesses in SHA-1 could be used to impersonate someone else via this attribute.

This would be difficult, though, as it would imply that the user we just authenticated was able to modify his/her own user identifier. Still, this would be possible under certain circumstances (e.g. security issues in the directory, or users with admin rights), so it's better if we avoid this possibility completely by using another hash function.

The problem here is that `eduPersonTargetedID` is used by many services to identify users univocally. If we modify how we generate the attribute, we need to allow the generation of both the old and new identifiers at the same time, and store them in different attributes. Given that this is an authproc filter, it should be possible to create the attribute with the new hash algorithm of choice, copy the resulting `eduPersonTargetedID` to a new attribute used while migrating, then regenerating that attribute again with the old SHA-1 hash.

I can envision this migration process with the inclusion of a new configuration option to specify the algorithm to use. That way, we make it possible to specify the algorithm (defaulting to SHA-1), and then users can configure the authproc with other algs, allowing for stronger hashes:

1. Generate the new `eduPersonTargetedID` with the new algorithm by configuring an authproc with the algorithm we want to use (i.e. SHA256).
2. Rename the `eduPersonTargetedID` we just generated with the help of another authproc filter with higher index than the previous (i.e. `core:AttributeCopy`, or `core:PHP`).
3. Generate the original `eduPersonTargetedID` with the `core:TargetedID` authproc filter with higher index than the previous, not specifying the hash algorithm.

@thijskh, @tvdijen comments? Thoughts?

**Comment by tvdijen** [comment]
I would opt for a deprecation notice on SHA-1 and have a safe default in the next release. Other than that, I have no remarks.

**Comment by trscavo** [comment]
FYI, there is a movement afoot to deprecate `eduPersonTargetedID`: https://spaces.internet2.edu/x/bI_TBg

**Comment by tvdijen** [comment]
Very interesting @trscavo!

**Comment by jaimeperez** [comment]
Thanks Tom! I was aware of the discussions, but didn't have time to catch up with them yet. It might well be that we don't even need to do this...

**Comment by pmeulen** [comment]
Regarding the security impact of using SHA-1 in the SSP NameID generation algorithm, I want to add a few observations:
- The currently known attack (SHAtter) requires the start of the plaintext (the prefix) to be under control by the attacker. It does not allow finding an collision for a typical existing SHA-1 hash.
- For generation of the NameID by SSP the plaintext being hashed includes a secret (see: https://github.com/simplesamlphp/simplesamlphp/blob/master/modules/core/lib/Auth/Process/TargetedID.php#L117-L121) at the beginning of the plaintext which should be unknown to the attacker. If it is known then other secrets from the configuration are probably exposed as well, like private key used to sign SAML assertions.
- Interestingly a SHA-1 collision generated using SHAtter can be detected based on the hashed value alone, also SHA-1 can be hardened against these attacks, both with little computational overhead: https://github.com/cr-marcstevens/sha1collisiondetection

My conclusion is that this means that the way the SHA-1 is used in the generation of NameID's in SSP is not impacted by the currently known attack methods (i.e. there is no better known way than a brute force attack). 

That being said. Attacks only get better, even if it is just because hardware is getting better. Thus it makes perfect sense to start thinking of a way to migrate from one NameID generation scheme to another. Which is not a easy problem to solve operationally, because it practically amounts to renumber user identifiers which requires changes at the SP as well. 

When we're talking about a standard for a new identifier as @trscavo mentions, can we also discuss (or even better standardise) a way of moving a user to a new identifier at the SP?

**Comment by jaimeperez** [comment]
Thanks @pmeulen! I fully agree. I wasn't worried about _SHAttered_ when I opened this issue, but mostly about any future issues that could arise. If I recall correctly, the _SHAttered_ attack even requires control of both preimages, so it's not enough to mount a second preimage attack.

In any case, I think it makes sense to start discussing a plan. It might be that there's no point on it if we stop to use _eduPersonTargetedID_ (that would resolve the issue by eliminating it), but I suspect if that happens it'll take a long time, so it's wise to have a plan to get away rather quickly from it, if needed.

Other than that, huge +1 to your request to discuss / standardise a way to _migrate_ identifiers at the service providers.

**Comment by trscavo** [comment]
> Other than that, huge +1 to your request to
> discuss / standardise a way to migrate identifiers
> at the service providers.

Wasn't the SAML2 Name Identifier Management Protocol invented precisely for this purpose?

**Comment by jaimeperez** [comment]
That's interesting. I always took that literally, as in, for `NameIDs`. Could it be used too to update the values of other attributes, such as `eduPersonTargetedID`?

**Comment by trscavo** [comment]
Yes, the SAML2 Name Identifier Management Protocol is designed for NameIDs (as its name implies) but there's a one-to-one correspondence between NameIDs and SAML attributes. See the MACE-Dir SAML Attribute Profiles for an explicit example.
